File: mempill-python/python/mempill/quickstart.py
"""
mempill five-minute quickstart.

Run with:
    python -m mempill.quickstart
    uv run python -m mempill.quickstart

Demonstrates the Tier-1 ergonomic API (remember / recall):
  1. Remember two facts about the same subject (city, with non-overlapping time windows).
  2. Recall — the open-ended Munich fact wins (most recent valid period).
  3. Contested — two timeless Functional facts on the same subject-line, showing
     that value=None + is_contested()=True + candidates populated is the correct
     read, NOT a NoBelief misread.

Zero imports of internal types (ProvenanceLabel, Disposition, ConfidenceDict, etc.).
"""

from mempill import open_in_memory, remember, recall, RememberOptions


def main() -> None:
    engine = open_in_memory()
    agent = "my-agent"
    remember(engine, agent, "user", "city", "Berlin",
        RememberOptions(valid_from="2020-01-01", valid_until="2025-01-01"))
    remember(engine, agent, "user", "city", "Munich",
        RememberOptions(valid_from="2025-01-01"))
    result = recall(engine, agent, "user", "city")
    assert result.as_str() == "Munich"
    assert not result.is_contested()
    remember(engine, agent, "acme", "ceo", "Alice")
    remember(engine, agent, "acme", "ceo", "Bob")
    ceo = recall(engine, agent, "acme", "ceo")
    assert ceo.is_contested()
    assert len(ceo.candidates) == 2
    assert ceo.value is None
    print("quickstart passed")


if __name__ == "__main__":
    main()

File: mempill-facade/examples/quickstart.rs
//! # quickstart — 5-minute mempill introduction
//!
//! Demonstrates the Tier-1 ergonomic API (`remember` / `recall`).
//! Zero internal-type imports required.
//!
//! Run with:
//!   cargo run -p mempill --example quickstart

use mempill::{open_default_in_memory, remember, recall, RememberOptions};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let engine = open_default_in_memory()?;
    let agent  = "my-agent";
    remember(&engine, agent, "user", "city", "Berlin",
        RememberOptions::new().valid_from("2020-01-01").valid_until("2025-01-01")).await?;
    remember(&engine, agent, "user", "city", "Munich",
        RememberOptions::new().valid_from("2025-01-01")).await?;
    let result = recall(&engine, agent, "user", "city").await?;
    assert_eq!(result.as_str(), Some("Munich"));
    assert!(!result.is_contested());
    remember(&engine, agent, "acme", "ceo", "Alice", RememberOptions::new()).await?;
    remember(&engine, agent, "acme", "ceo", "Bob",   RememberOptions::new()).await?;
    let ceo = recall(&engine, agent, "acme", "ceo").await?;
    assert!(ceo.is_contested());
    assert_eq!(ceo.candidates.len(), 2);
    assert!(ceo.value.is_none());
    println!("quickstart passed");
    Ok(())
}
